### PR TITLE
Add correctness test for video.

### PR DESCRIPTION
**Pull request by lina128**
This PR adds correctness test for golden video file.
1. [Video source](https://www.pexels.com/video/woman-doing-squats-4838220/): Download the 720 x 1280 resolution

2. Use ffmpeg to compress the video to 5fps with some loss to make the file smaller.
`ffmpeg -i original_pose.mp4 -r 5 -vcodec libx264 -crf 28 -profile:v baseline pose_squats.mp4`

3. For the 13sec video with 5fps, ran estimation on 65 frames. When testing, simulate a 30FPS video by overriding the timestamp.

4. The test also draws keypoints and skeleton of expected and actual. Video can be obtained from Browserstack.

This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/644)

### CODE REVIEW

File: pose-detection/src/blazepose/blazepose_test.ts [hunk 1]
```diff
@@ -22,9 +22,9 @@ import {ALL_ENVS, BROWSER_ENVS, describeWithFlags} from '@tensorflow/tfjs-core/d
 import {expectArraysClose} from '@tensorflow/tfjs-core/dist/test_util';
 
 import * as poseDetection from '../index';
-import {loadImage} from '../test_util';
+import {KARMA_SERVER, loadImage, loadVideo} from '../test_util';
 
-const MODEL_TYPE = ['fullbody', 'upperbody'];
+const UPPERBODY_ONLY = [false];
```
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose/blazepose_test.ts:27]: Why don't you run the test for upper-body?

File: pose-detection/src/blazepose/blazepose_test.ts [hunk 2]
```diff
@@ -135,3 +133,65 @@ describeWithFlags('Blazepose static image ', BROWSER_ENVS, () => {
...
+
+      let count = 0;
+      for (let i = 0; i < result.length; i++) {
+        const frameResult = result[i];
+        for (let j = 0; j < frameResult.length; j++) {
+          const keypoint = frameResult[j];
+          const x = keypoint[0];
+          const y = keypoint[1];
+
+          const ex = expectedFull[i][j][0];
```
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose/blazepose_test.ts:181]: I guess you want to take either expectedFull or expectedUpper (doesn't exist now) depending on the upperBodyOnly flag?

File: pose-detection/src/blazepose/blazepose_test.ts [hunk 3]
```diff
@@ -135,3 +133,65 @@ describeWithFlags('Blazepose static image ', BROWSER_ENVS, () => {
...
+          const ey = expectedFull[i][j][1];
+
+          if (Math.abs(x - ex) > 30 || Math.abs(y - ey) > 30) {
```
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose/blazepose_test.ts:184]: Please move `30` to some constant.

File: pose-detection/src/blazepose/blazepose_test.ts [hunk 4]
```diff
@@ -135,3 +133,65 @@ describeWithFlags('Blazepose static image ', BROWSER_ENVS, () => {
...
+
+      const result: number[][][] = [];
+
+      const callback = async(video: HTMLVideoElement): Promise<void> => {
+        const poses = await detector.estimatePoses(video);
+        result.push(poses[0].keypoints.map(kp => [kp.x, kp.y]));
+      };
+      await loadVideo('pose_squats.mp4', callback);
+
+      let count = 0;
```
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose/blazepose_test.ts:173]: `count` -> `errors_count`

File: pose-detection/src/blazepose/blazepose_test.ts [hunk 5]
```diff
@@ -135,3 +133,65 @@ describeWithFlags('Blazepose static image ', BROWSER_ENVS, () => {
...
+          const ey = expectedFull[i][j][1];
+
+          if (Math.abs(x - ex) > 30 || Math.abs(y - ey) > 30) {
+            count++;
+            console.log(`Frame ${i}, Keypoint ${j}: Actual [${x}, ${
+                y}], Expected [${ex}, ${ey}]`);
+          }
+        }
+      }
+      console.log(count);
```
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose/blazepose_test.ts:191]: Add some log message.

File: pose-detection/src/blazepose/blazepose_test.ts [hunk 6]
```diff
@@ -135,3 +133,65 @@ describeWithFlags('Blazepose static image ', BROWSER_ENVS, () => {
...
+      // expectArraysClose(result.slice(0, 65), expectedFull, 45);
```
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose/blazepose_test.ts:192]: Please uncomment and reuse the constant introduced above.
And increase the constant to 45 (or some smaller value if possible).

File: pose-detection/src/test_util.ts [hunk 7]
```diff
@@ -31,3 +33,43 @@ export async function loadImage(imagePath: string, width: number, height: number
...
+  const ctx = canvas.getContext('2d');
+
+  const promise = new Promise((resolve, reject) => {
+    video.onseeked = async () => {
+      canvas.height = video.videoHeight;
+      canvas.width = video.videoWidth;
+
+      ctx.drawImage(video, 200, 0);
+      await callback(video);
+      const nextTime = video.currentTime + 0.2;
```
Comment (ivan-grishchenko) [inline pose-detection/src/test_util.ts:56]: Why not same FPS as in Python test?

**Review by ivan-grishchenko — CHANGES_REQUESTED** [review]

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/644)* status: 0 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @lina128, @mhays-google, and @pyu10055)

___
*[pose-detection/src/test_util.ts, line 56 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5D-r1-56:-MY6CpVn5Pq6qZUOfLv2:bo722jw) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/test_util.ts#L56)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

Why not same FPS as in Python test?
</blockquote></details>

It is the same FPS, the Python test compress the video as 5 fps, and the result has 65 frames for the 13 second video.
___
*[pose-detection/src/blazepose/blazepose_test.ts, line 27 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5C-r1-27:-MY6D3k749Z-02TXu0TK:b-s6d8kl) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/blazepose/blazepose_test.ts#L27)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

Why don't you run the test for upper-body?
</blockquote></details>

This PR is still experimenting, once you reviewed and agreed with the approach, I will add upper-body to this list.
___
*[pose-detection/src/blazepose/blazepose_test.ts, line 181 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5C-r1-181:-MY6DKPqD6_OTECynern:b2nhswy) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/blazepose/blazepose_test.ts#L181)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

I guess you want to take either expectedFull or expectedUpper (doesn't exist now) depending on the upperBodyOnly flag?
</blockquote></details>

Yes

**Review by ivan-grishchenko — COMMENTED** [review]
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/644)* status: 0 of 1 approvals obtained (waiting on @bazarevsky, @lina128, @mhays-google, and @pyu10055)

___
*[pose-detection/src/test_util.ts, line 56 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5D-r1-56:-MY6QrdYB-XHzLhD6xTL:bn03c5a) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/test_util.ts#L56)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

It is the same FPS, the Python test compress the video as 5 fps, and the result has 65 frames for the 13 second video.
</blockquote></details>

As discussed during the call, this FPS is how frames are stored in a video, not how they are passed into the graph.
So this one can be any, while the other one should be 33.3333ms
___
*[pose-detection/src/blazepose/blazepose_test.ts, line 27 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5C-r1-27:-MY6RK3g6U-EDg3qEzK9:brxp9sq) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/blazepose/blazepose_test.ts#L27)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

This PR is still experimenting, once you reviewed and agreed with the approach, I will add upper-body to this list.
</blockquote></details>

Sounds good.
___
*[pose-detection/src/blazepose/blazepose_test.ts, line 181 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5C-r1-181:-MY6RW_eBZpMoEOZ7of-:b2pzwc8) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/blazepose/blazepose_test.ts#L181)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Yes
</blockquote></details>

I understand you'll add upper-body case later, but better to write code here in a way, that you won't forget to update it. I.e. check that upperBodyOnly is false for now or get expected value from expectedUpper constant which is define as NONE for now.

**Review by lina128 — COMMENTED** [review]
Hi Ivan, again, thank you for your thoughtful review! I've addressed all your comments, please take another look.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-:-MY7dYoIELzagDVV_q2X:bl9tuu1)* status: 0 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @lina128, @mhays-google, and @pyu10055)

___
*[pose-detection/src/test_util.ts, line 56 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5D-r1-56:-MY7bvDD-6LJ_kxCoWs5:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/test_util.ts#L56)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

As discussed during the call, this FPS is how frames are stored in a video, not how they are passed into the graph.
So this one can be any, while the other one should be 33.3333ms
</blockquote></details>

Done.
___
*[pose-detection/src/blazepose/blazepose_test.ts, line 27 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5C-r1-27:-MY7d8oy1yW1_UCn5khj:biulgh5) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/blazepose/blazepose_test.ts#L27)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

Sounds good.
</blockquote></details>

Added both upper body and full body tests.
___
*[pose-detection/src/blazepose/blazepose_test.ts, line 173 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5C-r1-173:-MY7dMdAD2W-etz143sY:b-6zrfd1) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/blazepose/blazepose_test.ts#L173)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

`count` -> `errors_count`
</blockquote></details>

Removed the debugging code.
___
*[pose-detection/src/blazepose/blazepose_test.ts, line 192 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MXuZlSm_lB9HrK1aS5C-r1-192:-MY7dWI63BWKw_4Fc8Ig:baiyqpa) ([raw file](https://github.com/tensorflow/tfjs-models/blob/2c833ee94844acfb329ef97f898928032aeaea00/pose-detection/src/blazepose/blazepose_test.ts#L192)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

Please uncomment and reuse the constant introduced above.
And increase the constant to 45 (or some smaller value if possible).
</blockquote></details>

Uncommented.

**Review by mhays-google — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/644)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @lina128, and @pyu10055)

___
*[pose-detection/demo/src/camera.js, line 112 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MYAvdOc8srlHMQzl1gf:-MYAvdOc8srlHMQzl1gg:b118u5s) ([raw file](https://github.com/tensorflow/tfjs-models/blob/446c34a0346663d74e74352407f8d77d37d7d28d/pose-detection/demo/src/camera.js#L112)):*
> ```JavaScript
>     this.ctx.lineWidth = params.DEFAULT_LINE_WIDTH;
> 
>     keypointInd.middle.forEach(i => this.drawKeypoint(keypoints[i]));
> ```

If middle is a typical iterable, Google style prohibits forEach and prefers "for i of middle".

**Review by ivan-grishchenko — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/644)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @lina128, @mhays-google, and @pyu10055)

___
*[pose-detection/src/test_util.ts, line 47 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MYAuGk94Z4UdAGPzOrK:-MYAuGk94Z4UdAGPzOrL:be9ux72) ([raw file](https://github.com/tensorflow/tfjs-models/blob/446c34a0346663d74e74352407f8d77d37d7d28d/pose-detection/src/test_util.ts#L47)):*
> ```XML
>   let simulatedTimestamp: number;
> 
>   const actualInterval = 1 / videoFPS;
> ```

Just to be sure. Result of this devision is always float? Not integer devision?
___
*[pose-detection/src/test_util.ts, line 65 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MYAuhC83anEfxZFUMTF:-MYAuhC83anEfxZFUMTG:bqnv8up) ([raw file](https://github.com/tensorflow/tfjs-models/blob/446c34a0346663d74e74352407f8d77d37d7d28d/pose-detection/src/test_util.ts#L65)):*
> ```XML
>         // We set the timestamp increment to 33.333 microseconds to simulate
>         // the 30 fps video input. We do this so that the filter uses the
>         // same fps as the python test.
> ```

I wouldn't mention python test here as this test will also be public, right?
Just some "reference test"
___
*[pose-detection/src/blazepose/blazepose_test.ts, line 29 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MYAv3IA2pYJjQ1ywCiV:-MYAv3IA2pYJjQ1ywCiW:b-go6ejh) ([raw file](https://github.com/tensorflow/tfjs-models/blob/446c34a0346663d74e74352407f8d77d37d7d28d/pose-detection/src/blazepose/blazepose_test.ts#L29)):*
> ```XML
> const UPPERBODY_ONLY = [false, true];
> const EPSILON_IMAGE = 10;
> const EPSILON_VIDEO = 50;
> ```

This epsilon is a little higher than I expected.
Would be nice for the test to also output rendering of actual vs. expected for the whole video to visually check that there is no major differences. And you can attach the resulting video to the bug as a proof.

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/644)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @mhays-google, and @pyu10055)

___
*[pose-detection/demo/src/camera.js, line 112 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MYAvdOc8srlHMQzl1gf:-MYBAVp53QgnM3CinSTO:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/446c34a0346663d74e74352407f8d77d37d7d28d/pose-detection/demo/src/camera.js#L112)):*
<details><summary><i>Previously, mhays-google (Michael Hays) wrote…</i></summary><blockquote>

If middle is a typical iterable, Google style prohibits forEach and prefers "for i of middle".
</blockquote></details>

Done.
___
*[pose-detection/src/test_util.ts, line 47 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MYAuGk94Z4UdAGPzOrK:-MYBAY3lA8-c9aIQwHA8:b-r4dob7) ([raw file](https://github.com/tensorflow/tfjs-models/blob/446c34a0346663d74e74352407f8d77d37d7d28d/pose-detection/src/test_util.ts#L47)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

Just to be sure. Result of this devision is always float? Not integer devision?
</blockquote></details>

Yes, always float.
___
*[pose-detection/src/test_util.ts, line 65 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MYAuhC83anEfxZFUMTF:-MYBAoop5q2FfCKg9-2R:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/446c34a0346663d74e74352407f8d77d37d7d28d/pose-detection/src/test_util.ts#L65)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

I wouldn't mention python test here as this test will also be public, right?
Just some "reference test"
</blockquote></details>

Done.
___
*[pose-detection/src/blazepose/blazepose_test.ts, line 29 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/644#-MYAv3IA2pYJjQ1ywCiV:-MYBkPXMBKY6R4irniV6:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/446c34a0346663d74e74352407f8d77d37d7d28d/pose-detection/src/blazepose/blazepose_test.ts#L29)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

This epsilon is a little higher than I expected.
Would be nice for the test to also output rendering of actual vs. expected for the whole video to visually check that there is no major differences. And you can attach the resulting video to the bug as a proof.
</blockquote></details>

Done.

**Review by ivan-grishchenko — APPROVED** [review]

**Pull request merged**